:sparkles: Combobox: Mulighet for å begrense hvor mange valg bruker kan ta

larseirikhansen opened the pull request:
### Description

I Combobox ønskes det mulighet for å begrense antall valgte valg, f.eks. "velg opptil tre favoritter fra listen".
Figma: https://www.figma.com/file/oy15hc0N3TfkE9Vu5FOyhZ/Combobox?type=design&node-id=0%3A1&mode=dev

### Change summary
- Essentially restructure the whole component

---
#### TODO
- [x] Add/update component-demos for component `aksel.nav.no/website/pages/eksempler`
- [x] Run minor `yarn changeset`

larseirikhansen committed:
yarn lock

larseirikhansen committed:
setting up maxSelectedOptions

larseirikhansen committed:
added correct margin

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox-limit-choices

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox-limit-choices

larseirikhansen committed:
Update tsconfig.build.json

it-vegard committed:
Make border of "max selected"-item in FilteredOptions overlap the list border

it-vegard committed:
Hide "add new value"-button if "max selected"-limit is reached

it-vegard committed:
Make MaxSelected example a bit more interactive

larseirikhansen committed:
wip styling max-selected option

larseirikhansen committed:
merge branch

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox-limit-choices' into combobox-limit-choices

it-vegard committed:
Merge branch 'main' into combobox-limit-choices

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox-limit-choices

it-vegard committed:
Merge branch 'main' into combobox-limit-choices

# Conflicts:
#	@navikt/core/react/src/form/combobox/FilteredOptions/FilteredOptions.tsx

it-vegard committed:
Disable filteredOptions that are not selected when maxSelectedOptions is reached

it-vegard committed:
Fix scrolling/clipping content in FilteredOptions

it-vegard committed:
Remove hover-effects from non-selectable options

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox-limit-choices

larseirikhansen committed:
add prop so user can customize message shown when max num of choices reached

larseirikhansen committed:
merge origin combobox-limit-choices

it-vegard committed:
Operate directly on the DOM nodes when moving virtual focus, instead of relying on index state.

The index state became stale/non-synced when replacing elements in the list.

it-vegard committed:
Merge branch 'main' into combobox-limit-choices

larseirikhansen committed:
make maxSelectedMessage static

it-vegard commented on @navikt/core/css/form/combobox.css:
```diff
@@ -268,6 +295,15 @@
+  background: var(--a-surface-neutral-subtle);
+}
+
+.navds-combobox__list-item:not(.navds-combobox__list-item__max-selected)[data-no-focus="true"] {
```
Denne treffer også "ingen søketreff" og "loading"-boksene. Bør spisse selectoren.

it-vegard commented on @navikt/core/css/form/combobox.css:
```diff
@@ -215,11 +215,22 @@
   padding: 0;
   box-shadow: var(--a-shadow-small);
   border-radius: var(--a-border-radius-medium);
-  gap: var(--a-spacing-1) 0;
```
Kan ta denne tilbake, og legge negativ margin-bottom på first-child

it-vegard commented on @navikt/core/css/form/combobox.css:
```diff
@@ -215,11 +215,22 @@
   color: var(--ac-combobox-list-text, var(--a-text-default));
 }
 
+.navds-combobox__list :where(li) {
```
Kan fjernes om gap tas tilbake som forklart over.

it-vegard commented on @navikt/core/css/form/combobox.css:
```diff
@@ -247,16 +256,34 @@
+}
+
+.navds-combobox__list-item__max-selected:hover {
+  background-color: unset;
```
```suggestion
```

larseirikhansen committed:
Remove max-selected:hover

Co-authored-by: Vegard Haugstvedt <it.vegard@gmail.com>

larseirikhansen committed:
yarn lock update

larseirikhansen committed:
'cursor: not allowed' only hits disabled options

larseirikhansen committed:
remove li-margins, add ul-gap with negative margin on max-selected-option

larseirikhansen committed:
remove 'a' from initialstate maxSelectedOptions Story

it-vegard commented on @navikt/core/css/form/combobox.css:
```diff
@@ -247,16 +245,30 @@
+
+.navds-combobox__list-item__max-selected {
+  background: var(--ac-combobox-list-item-max-selected-bg, var(--a-surface-info-subtle));
+  border-radius: 3px 3px 0 0;
```
```suggestion
  border-radius: 3px 3px 0 0; /* TODO: Change to default 4px border-radius when we refactor styling of dropdown */
```

it-vegard committed:
Update @navikt/core/css/form/combobox.css

it-vegard commented on @navikt/core/css/form/combobox.css:
```diff
@@ -268,6 +280,19 @@
+.navds-combobox__list-item:has(.navds-combobox__list-item__new-option)[data-no-focus="true"] {
+  background: var(--a-surface-neutral-subtle);
+}
+
```
```suggestion
```

it-vegard committed:
Update @navikt/core/css/form/combobox.css

it-vegard commented on @navikt/core/react/src/form/combobox/SelectedOptions/selectedOptionsContext.tsx:
```diff
@@ -12,11 +12,14 @@ import { ComboboxProps } from "../types";
   removeSelectedOption: (option: string) => void;
   prevSelectedOptions?: string[];
   selectedOptions: string[];
+  maxSelectedOptions?: number;
```
maxSelectedOptions -> numMaxSelected

Check if we can combine addNewValue, numMaxSelected++ in an option object

it-vegard replied:
maxSelected: {
  isLimitReached,
  limit,
  message
}

it-vegard replied:
Har implementert objekt-strukturen over.

it-vegard commented on @navikt/core/react/src/form/combobox/FilteredOptions/FilteredOptions.tsx:
```diff
@@ -41,9 +50,22 @@ const FilteredOptions = () => {
+        >
+          {maxSelectedMessage && maxSelectedOptions
+            ? maxSelectedMessage
+            : `${selectedOptions.length} av ${maxSelectedOptions} er valgt.`}
```
```suggestion
          {maxSelectedMessage ?? `${selectedOptions.length} av ${maxSelectedOptions} er valgt.`}
```

it-vegard committed:
Update @navikt/core/react/src/form/combobox/FilteredOptions/FilteredOptions.tsx

it-vegard commented on @navikt/core/react/src/form/combobox/FilteredOptions/FilteredOptions.tsx:
```diff
@@ -41,9 +50,20 @@ const FilteredOptions = () => {
       {isLoading && (
         <li
-          className="navds-combobox__list-item--loading"
+          className="navds-combobox__list-item navds-combobox__list-item__loading"
```
Sjekk hvorfor den er blitt `__loading`. Burde være `--loading` etter BEM

it-vegard replied:
Endret til --loading igjen

it-vegard commented on @navikt/core/react/src/form/combobox/FilteredOptions/useVirtualFocus.ts:
```diff
@@ -16,57 +15,85 @@ export type VirtualFocusType = {
+
+  const moveFocusDown = () => {
+    if (!activeElement) {
+      _moveFocusAndScrollTo(getElementsAbleToReceiveFocus()[0]);
```
Kanskje kall getElementsAbleToReceiveFocus én gang per metode -> lagre som en const

it-vegard replied:
Done

it-vegard commented on @navikt/core/react/src/form/combobox/FilteredOptions/useVirtualFocus.ts:
```diff
@@ -16,57 +15,85 @@ export type VirtualFocusType = {
-  const activeElement = elementsAbleToReceiveFocus[index];
+  const getListOfAllChildren = (): Array<HTMLElement> =>
+    containerRef?.children
+      ? Array.prototype.slice.call(containerRef?.children)
```
```suggestion
      ? Array.from(containerRef?.children)
```

it-vegard committed:
Update @navikt/core/react/src/form/combobox/FilteredOptions/useVirtualFocus.ts

it-vegard commented on @navikt/core/react/src/form/combobox/FilteredOptions/useVirtualFocus.ts:
```diff
@@ -16,57 +15,85 @@ export type VirtualFocusType = {
+  const getListOfAllChildren = (): Array<HTMLElement> =>
+    containerRef?.children
+      ? Array.from(containerRef?.children)
+      : [];
```
```suggestion
    Array.from(containerRef?.children ?? [])
```

it-vegard committed:
Update @navikt/core/react/src/form/combobox/FilteredOptions/useVirtualFocus.ts

it-vegard commented on @navikt/core/react/src/form/combobox/combobox.stories.tsx:
```diff
@@ -285,6 +285,35 @@ export const ComboboxSizes = () => (
+      label="Komboboks med begrenset antall valg"
+      options={options}
+      maxSelectedOptions={2}
+      maxSelectedMessage="2 av MAKS 2 er valgt."
```
```suggestion
```

it-vegard committed:
Update @navikt/core/react/src/form/combobox/combobox.stories.tsx

it-vegard commented on @navikt/core/react/src/form/combobox/types.ts:
```diff
@@ -97,6 +97,19 @@ export interface ComboboxProps
+   * @param max - The maximum number of selected options
+   * @returns The string / interpolated string to display
+   */
+  maxSelectedMessage?: string;
```
Oppdater JSDoc til at dette nå er en string. Endres når vi bytter til at disse ligger i et objekt.

it-vegard replied:
Er nå et objekt, med oppdatert JSDoc

larseirikhansen committed:
type assertion

larseirikhansen committed:
remove tab-index on wrapper div

it-vegard committed:
Merge branch 'main' into combobox-limit-choices

# Conflicts:
#	yarn.lock

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox-limit-choices' into combobox-limit-choices

larseirikhansen committed:
prototype for sticky non-selectable options

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox-limit-choices

larseirikhansen committed:
"yarn lock"

it-vegard committed:
Fix feil med at focus-indikator for input-feltet ikke ble vist

it-vegard committed:
Forenkle og fikse "scroll til element" i FilteredOptions

it-vegard committed:
Sørg for at disabled FilteredOptions ikke vises under "max antall"-meldingen

it-vegard committed:
Focused option should not be covered by "max selected" message

it-vegard committed:
Merge branch 'main' into combobox-limit-choices

# Conflicts:
#	@navikt/core/react/src/form/combobox/Combobox.tsx
#	@navikt/core/react/src/form/combobox/FilteredOptions/FilteredOptions.tsx
#	@navikt/core/react/src/form/combobox/FilteredOptions/useVirtualFocus.ts
#	@navikt/core/react/src/form/combobox/combobox.stories.tsx
#	yarn.lock

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox-limit-choices' into combobox-limit-choices

larseirikhansen committed:
yarn lock

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox-limit-choices

it-vegard committed:
Fix "scrollIntoView is not a function" error. Possibly needs to be checked for when it occurs, but I haven't figured it out yet

it-vegard committed:
Loading indicator is no longer exposed as an option, so we have to just search for it by the accessible name.

Should have been able to use findByRole("graphics-document", { name: "Søker..." }), but that is currently broken:
https://github.com/testing-library/dom-testing-library/issues/1131

it-vegard committed:
Aria-selected should not be placed on elements that are no longer role="option"

it-vegard committed:
Merge branch 'combobox-limit-choices' of https://github.com/navikt/aksel into combobox-limit-choices

# Conflicts:
#	@navikt/core/react/src/form/combobox/FilteredOptions/FilteredOptions.tsx
#	yarn.lock

it-vegard committed:
Merge branch 'main' into combobox-limit-choices

it-vegard committed:
Combine all "max selected options"-related props in one object

it-vegard committed:
Proper BEM syntax should use -- instead of __ for state-like classes, like --loading, --max-selected and --new-option

it-vegard committed:
Always show selected options in FilteredOptions, even if they don't match the search text

it-vegard committed:
Only to the getElementsAbleToReceiveFocus() call once per function, to reduce DOM lookups. Also changed isFocusOnThe[Top | Bottom] to functions, because it was needed for isFocusOnTheBottom to follow the same rule (and not store the value on the top level in the hook.

it-vegard committed:
Pressing "End" should not only move focus to the end of the list, but also open the list if it is closed. To mirror this, we close the list if moving focus back to the Input with "Home".

it-vegard committed:
Legg til Aksel-eksempel

it-vegard committed:
Merge branch 'main' into combobox-limit-choices

larseirikhansen committed:
Merge branch 'main' into combobox-limit-choices

larseirikhansen committed:
yarn changeset

it-vegard commented on @navikt/core/css/form/combobox.css:
```diff
@@ -241,34 +241,80 @@
+  border: 1px solid var(--ac-combobox-list-border-color, var(--a-border-divider));
+  box-shadow: var(--a-shadow-small);
 }
+*/
```
```suggestion
```

it-vegard committed:
Update @navikt/core/css/form/combobox.css

larseirikhansen committed:
change 'maxSelectedOptions' prop to 'maxSelected'

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox-limit-choices' into combobox-limit-choices

it-vegard committed:
Mark disabled options with aria-disabled

it-vegard committed:
Add max selected text to aria-describedby, so it is announced when the user enters the input field

it-vegard committed:
Add role="status" to the wrapper for "no hits", "loading" and "max selected" info messages, so they are announced when they appear

it-vegard committed:
Remove aria-selected (again)

it-vegard committed:
Merge branch 'main' into combobox-limit-choices

KenAJoh committed:
Merge branch 'main' into combobox-limit-choices

KenAJoh commented on @navikt/core/css/form/combobox.css:
```diff
@@ -241,34 +241,73 @@
+  left: 0;
+  right: 0;
+  z-index: 1;
+  gap: inherit;
```
Trengs disse? Merket ingen forskjell når jeg fjernet de på max-antall/loading 🤔 
```
  gap: inherit;
  background-color: inherit;
  display: inherit;
  flex-direction: inherit;
```

KenAJoh commented on @navikt/core/css/form/combobox.css:
```diff
@@ -241,34 +241,73 @@
+
+.navds-combobox__list-item--max-selected {
+  background: var(--ac-combobox-list-item-max-selected-bg, var(--a-surface-info-subtle));
+  border-radius: 3px 3px 0 0; /* TODO: Change to default 4px border-radius when we refactor styling of dropdown */
```
```suggestion
border-start-start-radius: calc(var(--a-border-radius-medium) - 1px);
border-start-end-radius: calc(var(--a-border-radius-medium) - 1px);
```

KenAJoh commented on @navikt/core/css/form/combobox.css:
```diff
@@ -280,6 +319,15 @@
+    .navds-combobox__list-item--max-selected,
+    .navds-combobox__list-item--loading,
+    .navds-combobox__list-item--no-options
+  )[data-no-focus="true"] {
```
```suggestion
.navds-combobox__list-item[data-no-focus="true"] {
```

HalvorHaugan replied:
Kan denne forenkles til dette? `.navds-combobox__list-options .navds-combobox__list-item[data-no-focus="true"]`

it-vegard replied:
Fjernet `:not` lokalt

KenAJoh commented on @navikt/core/react/src/form/combobox/ComboboxWrapper.tsx:
```diff
@@ -47,7 +47,7 @@ const ComboboxWrapper = ({
       className={cl(
         className,
-        "navds-form-field",
+        "navds-combobox-form-field",
```
Endringen fjernet spacing mellom label, input og error + "dot" før errormelding

Etter:
![Screenshot 2024-01-25 at 09 55 12](https://github.com/navikt/aksel/assets/26967723/68699e53-cd57-403e-8463-cdd4dcc58d5e)

Før:
![Screenshot 2024-01-25 at 09 55 43](https://github.com/navikt/aksel/assets/26967723/cc1b3d6d-0306-458e-a493-403c85f78724)

larseirikhansen replied:
fikset i [54e0e9e](https://github.com/navikt/aksel/pull/2260/commits/54e0e9e2684f2031b1a642db05c153997fe32c9b)

KenAJoh commented on @navikt/core/react/src/form/combobox/FilteredOptions/FilteredOptions.tsx:
```diff
@@ -27,111 +27,138 @@ const FilteredOptions = () => {
+              }
+              toggleOption(option, event);
+              if (!isMultiSelect && !selectedOptions.includes(option))
+                toggleIsListOpen(false);
```
```suggestion
              if (!isMultiSelect && !selectedOptions.includes(option)){
                              toggleIsListOpen(false);
              }
```

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox-limit-choices' into combobox-limit-choices

larseirikhansen committed:
fix: form error not showing correctly

it-vegard commented on @navikt/core/css/form/combobox.css:
```diff
@@ -241,34 +241,73 @@
+.navds-combobox__list--with-hover .navds-combobox__list-item:not([data-no-focus="true"]):hover {
   background-color: var(--ac-combobox-list-item-hover-bg, var(--a-surface-hover));
   cursor: pointer;
   border-left: 4px solid var(--ac-combobox-list-item-hover-border-left, var(--a-border-strong));
```
Denne må ignoreres for "add new". Finn workaround.

larseirikhansen replied:
Fikset i commit 08a6833

HalvorHaugan commented on aksel.nav.no/website/pages/eksempler/combobox/with-max-selected-limit.tsx:
```diff
@@ -0,0 +1,49 @@
+  return (
+    <div>
+      <UNSAFE_Combobox
+        label="Hva er de kuleste transportmidlene? (velg opp til 3)"
```
```suggestion
        label="Hva er de kuleste transportmidlene? (velg opptil 3)"
```

larseirikhansen committed:
fix: rm left hover-border on add-new-option item

it-vegard commented on @navikt/core/css/form/combobox.css:
```diff
@@ -1,3 +1,8 @@
+  display: flex;
+  flex-direction: column;
+}
+
```
```suggestion
```

it-vegard commented on @navikt/core/css/form/combobox.css:
```diff
@@ -241,34 +241,73 @@
+
+.navds-combobox__list-item--no-options:hover,
+.navds-combobox__list-item--loading:hover,
+.navds-combobox__list-item--max-selected:hover {
```
```suggestion
.navds-combobox__list_non-selectables:hover {
```

it-vegard committed:
Bare åpne listen som default, men la intern logikk styre etter første render

it-vegard committed:
Sticky trenger ikke ligger på hver enkelt av disse, da parent wrapperen også er sticky

it-vegard committed:
Combobox eksempler var ikke tilgjengelig i `yarn storybook:aksel` på grunn av manglende Demo-eksport

it-vegard committed:
Fjern ubrukt klasse som er i konflikt med navds-combobox__wrapper

it-vegard committed:
Trengs ikke etter at max selected ble flyttet ut av listen

it-vegard committed:
Denne koden gjør ingenting

it-vegard committed:
Trenger ikke bruke :not, fordi vi tok vekk data-no-focus i en commit tidligere i dag

it-vegard committed:
Bruk dynamiske selectedOptions for å kunne endre de

it-vegard committed:
Update @navikt/core/css/form/combobox.css

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

it-vegard committed:
Update @navikt/core/react/src/form/combobox/FilteredOptions/FilteredOptions.tsx

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

it-vegard committed:
Update aksel.nav.no/website/pages/eksempler/combobox/with-max-selected-limit.tsx

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

it-vegard committed:
Update @navikt/core/css/form/combobox.css

it-vegard committed:
Merge branch 'main' into combobox-limit-choices

it-vegard committed:
Do not render an empty ul for FilteredOptions, as it add a gap at the bottom of the container

it-vegard committed:
Do not render an empty non-selectables container, as it adds a gap at the top of the container

it-vegard committed:
Merge branch 'main' into combobox-limit-choices

# Conflicts:
#	yarn.lock

it-vegard committed:
Non-selectables should not have navds-combobox__list-item class, as it introduces a few issues with hovering, etc.

it-vegard committed:
Copy-pasted in an extra class that broke styling for max selected
